### PR TITLE
[MBL-1036] Block User Popup + Banner Message

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

1. Presents a confirmation alert when "Block this user" is selected from the block user action sheet
2. Presents a success or fail Message Banner when tapping block in this new alert. This isn't wired up to an API or anything yet. This behavior is hardcoded for now.

# 🤔 Why

Rather than sending a blocking request from the action sheet, we're going to use this confirmation alert pattern so that we can provide extra information and double-check that they want to block a user. 

# 🛠 How

The key areas are:
- Comments
- CommentReplies
- ProjectPage
- Messages

I'm taking advantage of the existing MessageBannerViewController that we use elsewhere in the app to display API response success and error banners. This means that each ViewController, except ProjectPage, needs to conform to `MessageBannerViewControllerPresenting` and present an instance of the banner when the corresponding ViewModel outputs a response from the API (which isn't wired up yet).

One caveat is that to keep users from spamming the block action sheet and alert, I'm disabling user interaction on the parent view until the success/error banner message is finished displaying. Once the success banner is dismissed on the project page, we auto-dismiss the view since the user doesn't want to see that content anymore. This could also set us up to more easily refresh our app content.
   - we won't dismiss when blocking from comments since users could block a user that doesn't own that project. 

So basically I've:
1. added functionality to display a new UIAlert when tapping the action sheet
2. when tapping Block on this new alert, we trigger an input to the ViewModel (this will eventually kick off our backend request) and then a new `userBlocked` output notifies the ViewController of which banner to display. Success or error.

# 👀 See

| 🦋 |
| --- |
Simulator Screen Recording - iPhone 15 Pro - 2023-11-07 at 10 44 02

# ✅ Acceptance criteria

_**Run the AppCenter Beta build. Make sure you're logged in and that the Block User FF is on.**_

- [x] Tapping the Block this user action sheet, from any of the key areas listed above, presents a UIAlert with an option to block or cancel
- [x] Tapping block user presents a banner message at the bottom of the screen. (this is temporarily controlled by a manual bool in the view model until we have an API to wire this up to)
- [x] If showing the success banner, the parent view is not tappable and dismisses once the banner hides
- [x] If showing the error banner, the parent view does not dismiss and users can try to again to block.

### CODE REVIEW

File: Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift:13]: Nit: it is more canonical to pass the calling object itself back in the delegate method. So the method would look like `messageBannerViewDidHide(view: MessageBannerViewController)` and then `bannerType` could be a `public private(set)`

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift:50]: I'm surprised to see that you need to manipulate `messageBannerViewController?.view` here. Two questions:

1. Why is this extra constraint necessary for this PR but not for other/older uses of `MessageBannerViewController`?
2. Can this be moved into a method on `MessageBannerViewController`? Something like `messageBannerViewController.fixTheBannerWidth(to: self.view.widthAnchor)` would be more self-documenting.

File: Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift:181]: I suspect that this behavior should be part of `MessageBannerViewController`. Having a banner that blocks interaction seems like a fairly common pattern. You could also add (if it doesn't already exist) a touch gesture that dismisses the banner when any additional area is tapped.

I peeked in `MessageBannerViewController` and it is an older-style class, which is probably way out of scope for any substantial edits, but it might be worthwhile folding in just this bit while you're in there.

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift:50]: So for some reason, that I haven't quite figured out yet, when this banner is added to. UITableViewController a right constraint of 1000 is added. If you look in `MessageBannerViewController` you'll see that it should be constrained to the right and left constraint of its parent view, but something is overwriting that. 
I like your idea of moving this to `MessageBannerViewController ` so that it can be more self-documenting. Without knowing the root cause I'm not sure how to approach that though.

File: Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift:181]: This is interesting. It would require some reworking of `MessageBannerViewController` to hold on to a reference to the parent view, but that might end up being less involved and we might be able to remove the delegate method altogether. I'll look into it.

a touch gesture that dismisses the banner already does exists 👍

File: Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift:13]: this method only needs to be aware of the banner type though. what value do we gain from passing the entire object?

File: Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift:181]: actually - since this banner is being used in other places in the app I'd rather address any refactoring of this class in a separate PR. This functionality is specific to the blocking flow currently as well and I'd rather not risk introducing regressions elsewhere.

File: Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift
Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift:13]: Ha, we gain nothing right now, but if you _did_ need the dismiss behavior to do something else in the future, it would be simpler to implement. However, that's why I'm thinking of this as a nit; it's really more of a style detail.

You may not need the delegate method at all if you're able to pull the `userInteractionDisabled` stuff into the view controller.

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift:50]: Very weird! (But intriguing). Have you tried the constraints/view debugger? Sometimes if you print the description of the constraint it'll tell you where it came from.

File: Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift
Comment (amy-at-kickstarter) [reply Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift:181]: I think that's fair.

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift]: I couldn't see why the tableviewcontroller was adding this extra trailing constraint to the banner but, I was able to get around this by setting the banner's widthAnchor rather than using leftAnchor + rightAnchor. 

I tested the other areas of the app where we use this banner and everything still looks good there.

File: Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift
Comment (amy-at-kickstarter) [inline Kickstarter-iOS/Features/Comments/Controllers/CommentRepliesViewController.swift:260]: I took a peek in `MessageBannerViewController`, and it does already reference `self.view.superview`, so it might be safe to move `isUserInteractionEnabled` in there. But that's not a blocker, as previously discussed.

**Review by amy-at-kickstarter — APPROVED** [review]
I didn't test on my machine, but 👀 code LGTM.

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (ifosli) [inline Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift:104]: Nit: Add a TODO here and for the other banner messages to update them once translated strings are ready

File: Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift
Comment (ifosli) [inline Kickstarter-iOS/Features/Messages/Controller/MessagesViewController.swift:192]: Nit: I don't think `author` is the right name in this context - just `user` would be better

File: Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift
Comment (ifosli) [inline Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift:111]: Did you want the message banner class to be responsible for turning on and off user interaction? I'm okay with that (as long as that works for all current call sites), but then the corresponding lines of code in the individual classes need to be deleted.

**Review by ifosli — APPROVED** [review]
Looks good! I added some nits that would be nice to fix before merging, but I don't think any of them are complicated enough that they'll require another review pass.

File: Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift
Comment (scottkicks) [reply Kickstarter-iOS/Features/MessageBanner/Controller/MessageBannerViewController.swift:111]: Yep - It does work for the other call sites. Good catch. I thought I had gotten them all. Will remove it from the last 2 classes.

**Pull request merged**
